src/nuke/GradeBeautyLayer.cpp

```diff
@@ -0,0 +1,276 @@
+
+private:
+    float blackpoint[3]{0.0f, 0.0f, 0.0f};
+    float whitepoint[3]{1.0f, 1.0f, 1.0f};
+    float lift[3]{0.0f, 0.0f, 0.0f};
+    float gain[3]{1.0f, 1.0f, 1.0f};
+    float offset[3]{0.0f, 0.0f, 0.0f};
+    float multiply[3]{1.0f, 1.0f, 1.0f};
+    float gamma[3]{1.0f, 1.0f, 1.0f};
+    bool reverse{false};
+    bool clampBlack{true};
+    bool clampWhite{false};
+    ChannelSet m_targetLayer{Mask_RGB};
+    ChannelSet m_sourceLayer{Mask_None};
+    ChannelSet m_selectedLayers;
+
+    // intermediate grade algorithm storage
+    float A[3]{0.0f, 0.0f, 0.0f};
+    float B[3]{0.0f, 0.0f, 0.0f};
+    float G[3]{0.0f, 0.0f, 0.0f};
+
+public:
+    void knobs(Knob_Callback);
+    void _validate(bool for_real);
+    bool pass_transform() const {return true;}
+    void in_channels(int, ChannelSet& channels) const;
+    void pixel_engine(const Row&, int, int, int, ChannelMask, Row&);
+    int knob_changed(Knob*);
+    const char* Class() const {return description.name;}
+    const char* node_help() const {return HELP;}
+    static const Iop::Description description;
+    // channel set that contains all channels that are modified by the node
+    ChannelSet activeChannelSet() const;
+    // pixel engine function when anything but the target layer is requested to render
+    void channelPixelEngine(const Row&, int, int, int, ChannelSet&, Row&);
+    // pixel engine functon when the target layer is requested to render
+    void beautyPixelEngine(const Row&, int y, int x, int r, ChannelSet&, Row&);
+    // This function calculates and stores the grade algorithm's intermediate calculations
+    bool precomputeValues();
+    GradeBeautyLayer(Node* node);
+    ~GradeBeautyLayer();
+};
+
+GradeBeautyLayer::GradeBeautyLayer(Node* node) : PixelIop(node)
+{
+    precomputeValues();
+}
+
+static Op* build(Node* node)
+{
+    return (new NukeWrapper(new GradeBeautyLayer(node)))->noChannels()->mixLuminance();
+}
+
+GradeBeautyLayer::~GradeBeautyLayer() {}
+
+const Iop::Description GradeBeautyLayer::description(
+    "GradeBeautyLayer",
+    "LayerAlchemy/GradeBeautyLayer",
+    build
+);
+
+ChannelSet GradeBeautyLayer::activeChannelSet() const
+{
+    ChannelSet outChans = ChannelSet(m_targetLayer + m_sourceLayer);
+    return outChans;
+}
+
+void GradeBeautyLayer::in_channels(int input_number, ChannelSet& mask) const {
+    mask += activeChannelSet();
+}
+
+bool GradeBeautyLayer::precomputeValues() {
+    bool changeZero = false;
+    for (unsigned int chanIdx = 0; chanIdx < 3; chanIdx++) {
+        float a = whitepoint[chanIdx] - blackpoint[chanIdx];
+        a = a ? (gain[chanIdx] - lift[chanIdx]) / a : 10000.0f;
+        a *= multiply[chanIdx];
+        float b = offset[chanIdx] + lift[chanIdx] - blackpoint[chanIdx] * a;
+        float g = LayerAlchemy::Utilities::validateGammaValue(gamma[chanIdx]);
+        A[chanIdx] = a;
+        B[chanIdx] = b;
+        G[chanIdx] = g;
+        if (a != 1.0f || b || g != 1.0f)
```

Comment:
Careful with the floating point comparison, it's easy to fall in the epsilon trap.
https://www.geeksforgeeks.org/comparison-float-value-c/